lang/java/reef-runtime-standalone/src/main/java/org/apache/reef/runtime/standalone/driver/NodeListManager.java

```diff
@@ -0,0 +1,304 @@
+import org.apache.reef.runtime.common.driver.resourcemanager.RuntimeStatusEventImpl;
+import org.apache.reef.runtime.common.files.FileResource;
+import org.apache.reef.runtime.common.files.REEFFileNames;
+import org.apache.reef.runtime.common.parameters.JVMHeapSlack;
+import org.apache.reef.runtime.common.utils.RemoteManager;
+import org.apache.reef.runtime.local.client.parameters.RootFolder;
+import org.apache.reef.runtime.yarn.driver.REEFEventHandlers;
+import org.apache.reef.runtime.standalone.client.parameters.NodeFolder;
+import org.apache.reef.runtime.standalone.client.parameters.NodeInfoSet;
+import org.apache.reef.tang.annotations.Parameter;
+import org.apache.reef.tang.exceptions.BindException;
+import org.apache.reef.tang.formats.ConfigurationSerializer;
+import org.apache.reef.util.CollectionUtils;
+import org.apache.reef.util.Optional;
+import org.apache.reef.wake.EventHandler;
+import org.apache.reef.wake.remote.RemoteMessage;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Management module for remote nodes in standalone runtime.
+ */
+public final class NodeListManager {
+  private static final Logger LOG = Logger.getLogger(NodeListManager.class.getName());
+
+  /**
+   * Map from containerID -> Container.
+   */
+  private final Map<String, Container> containers = new HashMap<>();
+
+  private final Set<String> nodeInfoSet;
+  private Iterator<String> nodeSetIterator;
+  private final ConfigurationSerializer configurationSerializer;
+  private final REEFFileNames fileNames;
+  private final double jvmHeapFactor;
+  private final REEFEventHandlers reefEventHandlers;
+  private final String errorHandlerRID;
+  private final String rootFolder;
+  private final String nodeFolder;
+  private final int sshPort = 22;
+
+  @Inject
+  NodeListManager(@Parameter(NodeInfoSet.class) final Set<String> nodeInfoSet,
+                  final ConfigurationSerializer configurationSerializer,
+                  final REEFFileNames fileNames,
+                  @Parameter(JVMHeapSlack.class) final double jvmHeapSlack,
+                  final RemoteManager remoteManager,
+                  final REEFEventHandlers reefEventHandlers,
+                  @Parameter(RootFolder.class) final String rootFolder,
+                  @Parameter(NodeFolder.class) final String nodeFolder) {
+    this.nodeInfoSet = nodeInfoSet;
+    this.configurationSerializer = configurationSerializer;
+    this.fileNames = fileNames;
+    this.jvmHeapFactor = 1.0 - jvmHeapSlack;
+    this.reefEventHandlers = reefEventHandlers;
+    this.errorHandlerRID = remoteManager.getMyIdentifier();
+    this.rootFolder = rootFolder;
+    this.nodeFolder = nodeFolder;
+
+    this.nodeSetIterator = this.nodeInfoSet.iterator();
+
+    remoteManager.registerHandler(ReefServiceProtos.RuntimeErrorProto.class,
```

Comment:
I am not sure what this piece of code does. Could you move this into a separate private method and leave a comment on what it does?